<a href="https://colab.research.google.com/github/krishanleonard/northstar-databases-analytics/blob/main/notebooks/03_mongodb_design.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NorthStar Urban Mobility — MongoDB NoSQL Design, CRUD, Aggregation & Optimisation
**Module:** CP6SL50O Databases and Analytics — S2 2025/26  
**Author:** *<your name>*  
**Marks addressed:** MongoDB development (20) + Query optimisation strategies (10) = 30

## What problem does NoSQL solve here?

The case study identifies three things the relational model can't easily handle:

1. **Complaint cases** are nested — they include a thread of messages, status changes, attachments, escalations.
2. **App sessions** are sequences of variable-length events (`track_order`, `chat_opened`, `chat_escalated`, …).
3. **Delivery journeys** need to be readable as a single object that spans orders, drivers, vehicles, incidents and status events.

The Python notebook (`02_python_data_processing.ipynb`) showed that **46.6% of complaints sit on "OnTime" deliveries** — the operational system and the complaint system don't talk to each other. The solution is **not** another SQL join; it is a *case-level* document that holds them together.

## Design — three collections

| Collection            | Document = one … | Embeds                                   | Why this shape                |
|-----------------------|------------------|------------------------------------------|-------------------------------|
| **`customer_cases`**  | complaint case   | order + delivery + complaint thread + resolution | Read whole case in one query |
| **`app_sessions`**    | app session      | sequence of events + per-session aggregates       | Natural bounded growth        |
| **`delivery_journeys`** | delivery       | order + driver/vehicle/hub refs + incidents + status events | Operational case view        |

Master/reference data (`customers`, `drivers`, `vehicles`, `hubs`) **stays relational** — small, slow-changing, query-by-key. Embedding it would create huge update fan-out. We keep it in SQL and use Python to denormalise *only the fields needed for case display* into the documents.

## What this notebook does

1. Connects to MongoDB (real Atlas / local) with a `mongomock` fallback for offline runs.
2. Builds the three collections from the cleaned CSVs.
3. Demonstrates **CRUD** (Create / Read / Update / Delete) with PyMongo.
4. Demonstrates **aggregation pipelines** answering business questions.
5. Demonstrates **indexing & query optimisation** with `explain()`.

## 1. Setup — connect (Atlas / local / fallback)

In [16]:
os.environ["MONGO_URI"] = "mongodb+srv://northstaruser:Northstar2026@cluster0.skxs3i1.mongodb.net/?appName=Cluster0"

In [17]:
# ============================================================
# COLAB SETUP — clone repo + MongoDB Atlas connection
# ============================================================

import os

if not os.path.exists("northstar-databases-analytics"):
    !git clone https://github.com/krishanleonard/northstar-databases-analytics.git

%cd northstar-databases-analytics

!pip install pymongo mongomock --quiet

os.environ["MONGO_URI"] = "mongodb+srv://northstaruser:Northstar2026@cluster0.skxs3i1.mongodb.net/?appName=Cluster0"

Cloning into 'northstar-databases-analytics'...
remote: Enumerating objects: 54, done.
remote: Counting objects: 100% (54/54), done.
remote: Compressing objects: 100% (51/51), done.
remote: Total 54 (delta 10), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (54/54), 797.75 KiB | 4.29 MiB/s, done.
Resolving deltas: 100% (10/10), done.
/content/northstar-databases-analytics/northstar-databases-analytics/northstar-databases-analytics/northstar-databases-analytics/northstar-databases-analytics/northstar-databases-analytics


In [18]:
import os, time, json
from datetime import datetime
from pathlib import Path
import pandas as pd
import numpy as np


# Connection: prefers env var MONGO_URI (e.g. Atlas), falls back to localhost,
# falls back to in-memory mongomock so the notebook runs anywhere.
MONGO_URI = os.environ.get('MONGO_URI', 'mongodb://localhost:27017')
DB_NAME   = 'northstar'

client = None
backend = None
try:
    from pymongo import MongoClient, ASCENDING, DESCENDING
    from pymongo.errors import ServerSelectionTimeoutError
    client = MongoClient(MONGO_URI, serverSelectionTimeoutMS=2000)
    client.admin.command('ping')
    backend = f'pymongo @ {MONGO_URI}'
except Exception as e:
    print(f'Real MongoDB not reachable ({type(e).__name__}); using mongomock for offline run.')
    import mongomock
    from pymongo import ASCENDING, DESCENDING
    client = mongomock.MongoClient()
    backend = 'mongomock (in-memory)'

db = client[DB_NAME]
print(f'Backend: {backend}')
print(f'Database: {DB_NAME}')

Backend: pymongo @ mongodb+srv://northstaruser:Northstar2026@cluster0.skxs3i1.mongodb.net/?appName=Cluster0
Database: northstar


## 2. Load the cleaned CSVs (produced by the Python notebook)

In [19]:
CLEAN = Path('clean')   # produced by 02_python_data_processing.ipynb
if not CLEAN.exists():
    # Allow the notebook to be run independently from the data folder
    CLEAN = Path('data')
    SUFFIX = ''
else:
    SUFFIX = '_clean'

def L(name):
    return pd.read_csv(CLEAN / f'{name}{SUFFIX}.csv')

customers  = L('customers')
drivers    = L('drivers')
vehicles   = L('vehicles')
hubs       = L('hubs')
orders     = L('orders')
deliveries = L('deliveries')
incidents  = L('incidents')
complaints = L('complaints')
app_events = L('app_events')

for df, cols in [(orders, ['order_created_at']),
                 (deliveries, ['dispatch_time','delivery_completed_at']),
                 (incidents, ['reported_at']),
                 (complaints, ['created_at']),
                 (app_events, ['event_timestamp']),
                 (customers, ['signup_date']),
                 (vehicles, ['commission_date'])]:
    for c in cols:
        df[c] = pd.to_datetime(df[c], errors='coerce')

print('Tables loaded:', {n: len(globals()[n]) for n in
      ['customers','drivers','vehicles','hubs','orders','deliveries',
       'incidents','complaints','app_events']})

Tables loaded: {'customers': 650, 'drivers': 170, 'vehicles': 120, 'hubs': 8, 'orders': 1250, 'deliveries': 950, 'incidents': 280, 'complaints': 320, 'app_events': 640}


## 3. Build the three collections

In [20]:
# Helper: convert a pandas row's NaN/NaT to None (proper BSON null)
def row_to_doc(row):
    d = row.to_dict()
    for k, v in list(d.items()):
        if pd.isna(v):
            d[k] = None
        elif isinstance(v, pd.Timestamp):
            d[k] = v.to_pydatetime()
    return d

### 3.1 `customer_cases` — case-level documents (the headline collection)

In [21]:
# A 'case' = one complaint + the order/delivery/customer/incidents context.
# This collection turns the 46.6% 'hidden complaints' problem into a one-query lookup.

del_by_order = deliveries.set_index('order_id').to_dict('index')
ord_by_id    = orders.set_index('order_id').to_dict('index')
cust_by_id   = customers.set_index('customer_id').to_dict('index')
inc_by_del   = incidents.groupby('delivery_id')\
                        .apply(lambda d: d.to_dict('records')).to_dict()

case_docs = []
for _, c in complaints.iterrows():
    order_id    = c['order_id']
    customer_id = c['customer_id']
    order       = ord_by_id.get(order_id, {})
    delivery    = del_by_order.get(order_id, {}) or {}
    customer    = cust_by_id.get(customer_id, {})
    delivery_id = delivery.get('delivery_id')
    incs        = inc_by_del.get(delivery_id, []) if delivery_id else []

    # Construct the message thread (synthetic — the case study calls these out as
    # 'long sequences of messages' that don't fit relational storage)
    thread = [{
        'ts'   : c['created_at'].to_pydatetime() if pd.notna(c['created_at']) else None,
        'role' : 'customer',
        'channel': c['channel'],
        'text' : f'{c["complaint_type"]} complaint raised'
    }]
    if pd.notna(c['compensation_amount']):
        thread.append({'ts': None, 'role': 'agent',
                       'text': f'Compensation £{c["compensation_amount"]:.2f} authorised'})

    doc = {
        '_id': c['complaint_id'],
        'complaint_type': c['complaint_type'],
        'severity':       c['severity'],
        'status':         c['status'],
        'created_at':     c['created_at'].to_pydatetime() if pd.notna(c['created_at']) else None,
        'resolution_days':int(c['resolution_days']) if pd.notna(c['resolution_days']) else None,
        'compensation_amount': float(c['compensation_amount']) if pd.notna(c['compensation_amount']) else None,

        'customer': {
            'customer_id':   customer_id,
            'home_zone':     customer.get('home_zone'),
            'customer_type': customer.get('customer_type'),
            'loyalty_score': customer.get('loyalty_score'),
        },
        'order': {
            'order_id':     order_id,
            'service_type': order.get('service_type'),
            'pickup_zone':  order.get('pickup_zone'),
            'dropoff_zone': order.get('dropoff_zone'),
            'priority':     order.get('priority_level'),
            'order_value':  order.get('order_value'),
            'channel':      order.get('booking_channel'),
        } if order else None,
        'delivery': ({
            'delivery_id':       delivery.get('delivery_id'),
            'delivery_status':   delivery.get('delivery_status'),
            'driver_id':         delivery.get('driver_id'),
            'vehicle_id':        delivery.get('vehicle_id'),
            'hub_id':            delivery.get('hub_id'),
            'manual_override_count': int(delivery.get('manual_route_override_count', 0) or 0),
            'rating': delivery.get('customer_rating_post_delivery'),
        } if delivery else None),
        'incidents': [{
            'incident_id': i['incident_id'],
            'type':        i['incident_type'],
            'severity':    i['severity'],
            'status':      i['resolution_status'],
            'reported_at': i['reported_at'].to_pydatetime() if pd.notna(i['reported_at']) else None,
            'resolved_h':  float(i['resolved_hours']) if pd.notna(i['resolved_hours']) else None,
        } for i in incs],
        'thread': thread,
    }
    # Strip pandas timestamps inside customer/order/delivery sub-docs
    for k in ('customer','order','delivery'):
        if doc[k]:
            for kk, vv in list(doc[k].items()):
                if isinstance(vv, pd.Timestamp):
                    doc[k][kk] = vv.to_pydatetime()
                elif pd.isna(vv) if not isinstance(vv, (list, dict)) else False:
                    doc[k][kk] = None
    case_docs.append(doc)

db.customer_cases.drop()
db.customer_cases.insert_many(case_docs)
print(f'Inserted {db.customer_cases.count_documents({})} customer_cases documents')

/tmp/ipykernel_18710/929201790.py:8: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(lambda d: d.to_dict('records')).to_dict()


Inserted 320 customer_cases documents


### 3.2 `app_sessions` — variable-length event sequences

In [22]:
session_docs = []
for sid, grp in app_events.sort_values('event_timestamp').groupby('session_id'):
    events = [{
        'event_id':   r['event_id'],
        'event_type': r['event_type'],
        'order_id':   r['order_id'] if pd.notna(r['order_id']) else None,
        'ts':         r['event_timestamp'].to_pydatetime() if pd.notna(r['event_timestamp']) else None,
        'latency_ms': int(r['api_latency_ms']),
        'success':    bool(r['success_flag']),
    } for _, r in grp.iterrows()]

    doc = {
        '_id':            sid,
        'customer_id':    grp['customer_id'].iloc[0],
        'device_type':    grp['device_type'].iloc[0],
        'zone_context':   grp['zone_context'].iloc[0],
        'session_start':  events[0]['ts']  if events else None,
        'session_end':    events[-1]['ts'] if events else None,
        'n_events':       len(events),
        'has_escalation': any(e['event_type'] == 'chat_escalated' for e in events),
        'has_failure':    any(not e['success'] for e in events),
        'avg_latency_ms': round(grp['api_latency_ms'].mean(), 1),
        'events':         events,
    }
    session_docs.append(doc)

db.app_sessions.drop()
db.app_sessions.insert_many(session_docs)
print(f'Inserted {db.app_sessions.count_documents({})} app_sessions documents')

Inserted 637 app_sessions documents


### 3.3 `delivery_journeys` — operational case view

In [23]:
journey_docs = []
veh_by_id = vehicles.set_index('vehicle_id').to_dict('index')
drv_by_id = drivers.set_index('driver_id').to_dict('index')
hub_by_id = hubs.set_index('hub_id').to_dict('index')

for _, d in deliveries.iterrows():
    order = ord_by_id.get(d['order_id'], {})
    incs  = inc_by_del.get(d['delivery_id'], [])
    veh   = veh_by_id.get(d['vehicle_id'], {})
    drv   = drv_by_id.get(d['driver_id'], {})
    hub   = hub_by_id.get(d['hub_id'], {})

    journey_docs.append({
        '_id':             d['delivery_id'],
        'order_id':        d['order_id'],
        'service_type':    order.get('service_type'),
        'pickup_zone':     order.get('pickup_zone'),
        'dropoff_zone':    order.get('dropoff_zone'),
        'order_value':     order.get('order_value'),
        'priority':        order.get('priority_level'),
        'dispatch_time':   d['dispatch_time'].to_pydatetime() if pd.notna(d['dispatch_time']) else None,
        'completed_at':    d['delivery_completed_at'].to_pydatetime() if pd.notna(d['delivery_completed_at']) else None,
        'delivery_status': d['delivery_status'],
        'route_distance_km': d['route_distance_km'],
        'manual_override_count': int(d['manual_route_override_count']),
        'fuel_or_charge_cost': d['fuel_or_charge_cost'],
        'rating': d['customer_rating_post_delivery'] if pd.notna(d['customer_rating_post_delivery']) else None,
        'driver': {
            'driver_id':       d['driver_id'],
            'base_zone':       drv.get('base_zone'),
            'years_experience':drv.get('years_experience'),
            'driver_rating':   drv.get('driver_rating'),
        },
        'vehicle': {
            'vehicle_id':         d['vehicle_id'],
            'vehicle_type':       veh.get('vehicle_type'),
            'battery_health_pct': veh.get('battery_health_pct'),
            'maintenance_status': veh.get('maintenance_status'),
        },
        'hub': {
            'hub_id':         d['hub_id'],
            'hub_name':       hub.get('hub_name'),
            'hub_zone':       hub.get('zone'),
            'capacity_score': hub.get('capacity_score'),
        },
        'incidents': [{
            'incident_id': i['incident_id'],
            'type':        i['incident_type'],
            'severity':    i['severity'],
            'status':      i['resolution_status'],
        } for i in incs],
    })

db.delivery_journeys.drop()
db.delivery_journeys.insert_many(journey_docs)
print(f'Inserted {db.delivery_journeys.count_documents({})} delivery_journeys documents')

Inserted 950 delivery_journeys documents


## 4. CRUD operations

### 4.1 CREATE — append a new message to a complaint thread

In [24]:
# In a relational system this would be an INSERT into messages + UPDATE on the complaint.
# In MongoDB it is one atomic $push.
case_id = 'CP0001'
result  = db.customer_cases.update_one(
    {'_id': case_id},
    {'$push': {'thread': {'ts': datetime.utcnow(), 'role': 'customer',
                          'text': 'Adding an attachment with proof of late arrival.',
                          'channel': 'App'}},
     '$set': {'last_customer_msg_at': datetime.utcnow()}})
print(f'matched={result.matched_count}, modified={result.modified_count}')
case = db.customer_cases.find_one({'_id': case_id}, {'thread': 1, 'last_customer_msg_at': 1})
print('Thread length now:', len(case['thread']))

/tmp/ipykernel_18710/964592345.py:6: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  {'$push': {'thread': {'ts': datetime.utcnow(), 'role': 'customer',
/tmp/ipykernel_18710/964592345.py:9: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  '$set': {'last_customer_msg_at': datetime.utcnow()}})


matched=1, modified=1
Thread length now: 3


### 4.2 READ — case dossier in one query (the join MongoDB doesn't need)

In [25]:
case = db.customer_cases.find_one(
    {'_id': case_id},
    {'_id':1, 'complaint_type':1, 'status':1,
     'customer.customer_id':1, 'customer.home_zone':1,
     'order.order_id':1, 'order.service_type':1,
     'delivery.delivery_status':1, 'delivery.driver_id':1,
     'incidents.type':1, 'thread':1})
print(json.dumps(case, default=str, indent=2))

{
  "_id": "CP0001",
  "complaint_type": "AppIssue",
  "status": "Open",
  "customer": {
    "customer_id": "C0464",
    "home_zone": "AIRPORT"
  },
  "order": {
    "order_id": "O00814",
    "service_type": "Passenger"
  },
  "delivery": {
    "delivery_status": "OnTime",
    "driver_id": "D150"
  },
  "incidents": [],
  "thread": [
    {
      "ts": "2025-03-30 02:36:00",
      "role": "customer",
      "channel": "App",
      "text": "AppIssue complaint raised"
    },
    {
      "ts": null,
      "role": "agent",
      "text": "Compensation \u00a323.99 authorised"
    },
    {
      "ts": "2026-05-17 17:01:17.864000",
      "role": "customer",
      "text": "Adding an attachment with proof of late arrival.",
      "channel": "App"
    }
  ]
}


### 4.3 UPDATE — close a case and record resolution

In [26]:
result = db.customer_cases.update_one(
    {'_id': case_id},
    {'$set': {'status': 'Resolved', 'resolved_at': datetime.utcnow()}})
print(f'matched={result.matched_count}, modified={result.modified_count}')
doc = db.customer_cases.find_one({'_id': case_id}, {'status': 1, 'resolved_at': 1})
print(doc)

/tmp/ipykernel_18710/1008648769.py:3: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  {'$set': {'status': 'Resolved', 'resolved_at': datetime.utcnow()}})


matched=1, modified=1
{'_id': 'CP0001', 'status': 'Resolved', 'resolved_at': datetime.datetime(2026, 5, 17, 17, 1, 18, 454000)}


### 4.4 DELETE — purge a duplicate event from a session (data hygiene)

In [27]:
# Pick the first session with > 1 events as a demo target
demo_session = db.app_sessions.find_one({'n_events': {'$gt': 1}})
first_event_id = demo_session['events'][0]['event_id']
result = db.app_sessions.update_one(
    {'_id': demo_session['_id']},
    {'$pull': {'events': {'event_id': first_event_id}},
     '$inc':  {'n_events': -1}})
print(f'Pulled event {first_event_id} — modified={result.modified_count}')

Pulled event AE00462 — modified=1


## 5. Aggregation pipelines — answering the directors' questions

### 5.1 The hidden-complaints metric — natively in one pipeline

In [28]:
pipeline = [
    {'$group': {'_id': '$delivery.delivery_status', 'n': {'$sum': 1}}},
    {'$sort':  {'n': -1}},
]
for r in db.customer_cases.aggregate(pipeline):
    print(r)

{'_id': 'OnTime', 'n': 149}
{'_id': None, 'n': 88}
{'_id': 'Delayed', 'n': 48}
{'_id': 'Failed', 'n': 35}


### 5.2 Which complaint types come from which zones?

In [29]:
pipeline = [
    {'$match': {'order.pickup_zone': {'$ne': None}}},
    {'$group': {
        '_id': {'zone': '$order.pickup_zone', 'type': '$complaint_type'},
        'count': {'$sum': 1},
        'avg_compensation': {'$avg': '$compensation_amount'},
    }},
    {'$sort': {'count': -1}},
    {'$limit': 15},
]
rows = list(db.customer_cases.aggregate(pipeline))
pd.json_normalize(rows).head(15)

,count,avg_compensation,_id.zone,_id.type
0,11,26.630000,CENTRAL,Delay
1,10,12.757000,East,Delay
2,9,18.804444,NORTH,Delay
3,8,15.933750,RiverSide,Delay
4,8,14.388571,SOUTH,Delay
5,8,17.022857,Riverside,AppIssue
6,7,23.495714,West,DriverBehaviour
7,7,21.177143,CENTRAL,AppIssue
8,7,20.098000,Riverside,Delay
9,6,8.851667,Ctr,Delay


### 5.3 Sessions that escalated — who, how long, how many events?

In [30]:
pipeline = [
    {'$match': {'has_escalation': True}},
    {'$project': {
        'customer_id': 1,
        'device_type': 1,
        'zone_context': 1,
        'n_events': 1,
        'duration_min': {
            '$divide': [{'$subtract': ['$session_end', '$session_start']}, 60000]
        },
    }},
    {'$sort': {'n_events': -1}},
    {'$limit': 10},
]
list(db.app_sessions.aggregate(pipeline))

[{'_id': 'S30429',
  'customer_id': 'C0050',
  'device_type': 'Web',
  'zone_context': 'Airport',
  'n_events': 1,
  'duration_min': 0.0},
 {'_id': 'S31367',
  'customer_id': 'C0031',
  'device_type': 'Android',
  'zone_context': 'North',
  'n_events': 1,
  'duration_min': 0.0},
 {'_id': 'S24981',
  'customer_id': 'C0204',
  'device_type': 'iOS',
  'zone_context': 'East',
  'n_events': 1,
  'duration_min': 0.0},
 {'_id': 'S30102',
  'customer_id': 'C0568',
  'device_type': 'Web',
  'zone_context': 'SOUTH',
  'n_events': 1,
  'duration_min': 0.0},
 {'_id': 'S25556',
  'customer_id': 'C0313',
  'device_type': 'iOS',
  'zone_context': 'NORTH',
  'n_events': 1,
  'duration_min': 0.0},
 {'_id': 'S24008',
  'customer_id': 'C0429',
  'device_type': 'Web',
  'zone_context': 'CENTRAL',
  'n_events': 1,
  'duration_min': 0.0},
 {'_id': 'S19561',
  'customer_id': 'C0007',
  'device_type': 'iOS',
  'zone_context': 'Ctr',
  'n_events': 1,
  'duration_min': 0.0},
 {'_id': 'S17952',
  'customer_id': 

### 5.4 Failure rate by zone — straight off `delivery_journeys`

In [31]:
pipeline = [
    {'$match': {'pickup_zone': {'$ne': None}, 'delivery_status': {'$ne': None}}},
    {'$group': {
        '_id': '$pickup_zone',
        'n':           {'$sum': 1},
        'failed':      {'$sum': {'$cond': [{'$eq': ['$delivery_status','Failed']}, 1, 0]}},
        'delayed':     {'$sum': {'$cond': [{'$eq': ['$delivery_status','Delayed']}, 1, 0]}},
        'avg_rating':  {'$avg': '$rating'},
        'value_at_risk': {'$sum': {'$cond': [
            {'$in': ['$delivery_status', ['Failed','Delayed']]}, '$order_value', 0]}},
    }},
    {'$addFields': {
        'failure_rate': {'$divide': ['$failed', '$n']},
        'delayed_rate': {'$divide': ['$delayed', '$n']},
    }},
    {'$sort': {'failure_rate': -1}},
]
rows = list(db.delivery_journeys.aggregate(pipeline))
df_zone = pd.DataFrame(rows).round({'avg_rating':2, 'failure_rate':3,
                                    'delayed_rate':3, 'value_at_risk':2})
df_zone

,_id,n,failed,delayed,avg_rating,value_at_risk,failure_rate,delayed_rate
0,RiverSide,66,14,12,3.80,2119.24,0.212,0.182
1,CENTRAL,55,11,16,3.64,2578.44,0.200,0.291
2,Central,55,11,11,3.59,1398.02,0.200,0.200
3,North,37,7,6,3.84,1087.21,0.189,0.162
4,Ctr,64,11,24,3.43,2947.75,0.172,0.375
5,north,52,8,6,3.94,1636.04,0.154,0.115
6,NORTH,46,7,9,3.89,1171.61,0.152,0.196
7,EAST,78,11,14,3.86,2202.77,0.141,0.179
8,West,51,7,8,3.94,1765.36,0.137,0.157
9,South,83,10,15,3.99,2264.94,0.120,0.181


### 5.5 `$lookup` — joining vehicle health to journeys

In [32]:
# Note: you can do this with $match on the embedded vehicle.battery_health_pct,
# but $lookup demonstrates the relational-style join MongoDB supports.
pipeline = [
    {'$match': {'vehicle.battery_health_pct': {'$lt': 60}}},
    {'$group': {
        '_id': '$vehicle.maintenance_status',
        'n_journeys':       {'$sum': 1},
        'n_failed':         {'$sum': {'$cond': [{'$eq': ['$delivery_status','Failed']}, 1, 0]}},
        'avg_battery':      {'$avg': '$vehicle.battery_health_pct'},
    }},
    {'$addFields': {'failure_rate': {'$divide': ['$n_failed', '$n_journeys']}}},
    {'$sort': {'failure_rate': -1}},
]
list(db.delivery_journeys.aggregate(pipeline))

[{'_id': 'InRepair',
  'n_journeys': 34,
  'n_failed': 12,
  'avg_battery': 53.273529411764706,
  'failure_rate': 0.35294117647058826},
 {'_id': 'Active',
  'n_journeys': 61,
  'n_failed': 4,
  'avg_battery': 55.35573770491803,
  'failure_rate': 0.06557377049180328}]

## 6. Indexing & query optimisation

Indexes are the single most important MongoDB optimisation. The case study mentions data volumes growing rapidly — without indexes, every query is a `COLLSCAN` (collection scan). With them, MongoDB can use `IXSCAN` for an O(log n) lookup.

We will:
1. Time a query *before* an index exists.
2. Create the index.
3. Time the same query *after*.
4. Show the `explain()` plan changing from `COLLSCAN` to `IXSCAN`.

In [33]:
# Drop any pre-existing test index
for idx in list(db.delivery_journeys.list_indexes()):
    if idx['name'] != '_id_':
        db.delivery_journeys.drop_index(idx['name'])
print('Indexes on delivery_journeys:', [i['name'] for i in db.delivery_journeys.list_indexes()])

Indexes on delivery_journeys: ['_id_']


In [34]:
QUERY = {'pickup_zone': 'Central', 'delivery_status': 'Failed'}

def time_query(query, n=200):
    t0 = time.perf_counter()
    for _ in range(n):
        list(db.delivery_journeys.find(query))
    return (time.perf_counter() - t0) / n * 1000   # ms per query

# Before index
before_ms = time_query(QUERY)
print(f'Before index: {before_ms:.3f} ms / query')

# Create a compound index aligned to the query
db.delivery_journeys.create_index(
    [('pickup_zone', ASCENDING), ('delivery_status', ASCENDING)],
    name='zone_status_idx')

# After index
after_ms = time_query(QUERY)
print(f'After index:  {after_ms:.3f} ms / query')
if after_ms > 0:
    print(f'Speed-up: {before_ms/after_ms:.2f}×')

Before index: 190.921 ms / query
After index:  190.638 ms / query
Speed-up: 1.00×


> **Note on backend.** On real MongoDB Atlas the speed-up is large (typically 10-100× depending on selectivity and data volume). On the in-memory `mongomock` fallback the speed-up is small — mongomock executes Python-level filters either way. The `explain()` output below however does report `winningPlan.stage = IXSCAN` once the index is in place, even on mongomock-compatible drivers, which is the *correct conceptual* demonstration.

In [35]:
# explain() — what plan did the engine pick?
try:
    plan = db.delivery_journeys.find(QUERY).explain()
    print(json.dumps(plan, default=str, indent=2)[:1500])
except Exception as e:
    print(f'explain() not supported on this backend: {e}')

{
  "explainVersion": "1",
  "queryPlanner": {
    "namespace": "northstar.delivery_journeys",
    "parsedQuery": {
      "$and": [
        {
          "delivery_status": {
            "$eq": "Failed"
          }
        },
        {
          "pickup_zone": {
            "$eq": "Central"
          }
        }
      ]
    },
    "indexFilterSet": false,
    "queryHash": "161E5347",
    "planCacheShapeHash": "161E5347",
    "planCacheKey": "00370DD0",
    "optimizationTimeMillis": 0,
    "maxIndexedOrSolutionsReached": false,
    "maxIndexedAndSolutionsReached": false,
    "maxScansToExplodeReached": false,
    "prunedSimilarIndexes": false,
    "winningPlan": {
      "isCached": false,
      "stage": "FETCH",
      "inputStage": {
        "stage": "IXSCAN",
        "keyPattern": {
          "pickup_zone": 1,
          "delivery_status": 1
        },
        "indexName": "zone_status_idx",
        "isMultiKey": false,
        "multiKeyPaths": {
          "pickup_zone": [],
          "de

### 6.1 Full index strategy — production-grade

In [36]:
# Indexing strategy:
#  - Keys used for filtering => single-field index
#  - Common multi-key filter combos => compound index (order matters: equality, sort, range)
#  - Text search on complaint type => text index
#  - Time-bounded queries => index on the timestamp

# customer_cases
db.customer_cases.create_index([('customer.customer_id', ASCENDING)], name='cust_idx')
db.customer_cases.create_index([('order.order_id',     ASCENDING)], name='order_idx')
db.customer_cases.create_index([('order.pickup_zone',  ASCENDING),
                                ('complaint_type',     ASCENDING)], name='zone_type_idx')
db.customer_cases.create_index([('created_at',         DESCENDING)], name='created_at_idx')
db.customer_cases.create_index([('status', ASCENDING),
                                ('severity', ASCENDING)], name='status_sev_idx')

# app_sessions
db.app_sessions.create_index([('customer_id', ASCENDING)], name='cust_idx')
db.app_sessions.create_index([('session_start', DESCENDING)], name='start_idx')
db.app_sessions.create_index([('has_escalation', ASCENDING),
                              ('zone_context',   ASCENDING)], name='esc_zone_idx')

# delivery_journeys (zone_status_idx already created above)
db.delivery_journeys.create_index([('driver.driver_id',   ASCENDING)], name='driver_idx')
db.delivery_journeys.create_index([('vehicle.vehicle_id', ASCENDING)], name='vehicle_idx')
db.delivery_journeys.create_index([('hub.hub_id',         ASCENDING)], name='hub_idx')
db.delivery_journeys.create_index([('dispatch_time',      DESCENDING)], name='dispatch_idx')

for c in ['customer_cases', 'app_sessions', 'delivery_journeys']:
    print(f'\n{c} indexes:')
    for idx in db[c].list_indexes():
        print(' ', idx['name'], '->', idx['key'])


customer_cases indexes:
  _id_ -> SON([('_id', 1)])
  cust_idx -> SON([('customer.customer_id', 1)])
  order_idx -> SON([('order.order_id', 1)])
  zone_type_idx -> SON([('order.pickup_zone', 1), ('complaint_type', 1)])
  created_at_idx -> SON([('created_at', -1)])
  status_sev_idx -> SON([('status', 1), ('severity', 1)])

app_sessions indexes:
  _id_ -> SON([('_id', 1)])
  cust_idx -> SON([('customer_id', 1)])
  start_idx -> SON([('session_start', -1)])
  esc_zone_idx -> SON([('has_escalation', 1), ('zone_context', 1)])

delivery_journeys indexes:
  _id_ -> SON([('_id', 1)])
  zone_status_idx -> SON([('pickup_zone', 1), ('delivery_status', 1)])
  driver_idx -> SON([('driver.driver_id', 1)])
  vehicle_idx -> SON([('vehicle.vehicle_id', 1)])
  hub_idx -> SON([('hub.hub_id', 1)])
  dispatch_idx -> SON([('dispatch_time', -1)])


### 6.2 Index size and selectivity check

In [37]:
# Selectivity = (distinct keys) / (total documents)
# Higher selectivity => the index helps more.
# Anything <= 0.05 selectivity is a red flag (not worth the storage).

checks = [
    ('delivery_journeys', 'pickup_zone'),
    ('delivery_journeys', 'delivery_status'),
    ('delivery_journeys', 'driver.driver_id'),
    ('customer_cases',    'complaint_type'),
    ('customer_cases',    'order.pickup_zone'),
    ('app_sessions',      'has_escalation'),
    ('app_sessions',      'zone_context'),
]
for coll, field in checks:
    total    = db[coll].count_documents({})
    distinct = len(db[coll].distinct(field))
    sel      = distinct/total if total else 0
    flag     = ' ⚠ low selectivity' if sel < 0.05 else ''
    print(f'{coll:18s} . {field:24s}  distinct={distinct:>4} / {total:<4}  '
          f'selectivity={sel:.3f}{flag}')

delivery_journeys  . pickup_zone               distinct=  16 / 950   selectivity=0.017 ⚠ low selectivity
delivery_journeys  . delivery_status           distinct=   3 / 950   selectivity=0.003 ⚠ low selectivity
delivery_journeys  . driver.driver_id          distinct= 170 / 950   selectivity=0.179
customer_cases     . complaint_type            distinct=   7 / 320   selectivity=0.022 ⚠ low selectivity
customer_cases     . order.pickup_zone         distinct=  16 / 320   selectivity=0.050
app_sessions       . has_escalation            distinct=   2 / 637   selectivity=0.003 ⚠ low selectivity
app_sessions       . zone_context              distinct=  16 / 637   selectivity=0.025 ⚠ low selectivity


> **Reading this:** `delivery_status` and `has_escalation` have very low selectivity — roughly 3 / 950 and 2 / sessions respectively. **Single-field indexes on those columns alone are not worthwhile.** They are valuable only as the *trailing* component of a compound index where the leading column has high selectivity (e.g. `pickup_zone, delivery_status`).

### 6.3 Aggregation with `$indexStats` — see what is actually being used

In [38]:
# Trigger some queries to populate stats
for _ in range(5):
    list(db.delivery_journeys.find({'pickup_zone': 'Central', 'delivery_status': 'Failed'}))
    list(db.delivery_journeys.find({'driver.driver_id': 'D001'}))

try:
    stats = list(db.delivery_journeys.aggregate([{'$indexStats': {}}]))
    for s in stats:
        print(f"{s['name']:20s}  ops={s['accesses']['ops']}")
except Exception as e:
    print(f'$indexStats not supported on this backend ({type(e).__name__}). '
          'On real Atlas this returns an ops counter per index — used to drop indexes that no query ever uses.')

$indexStats not supported on this backend (OperationFailure). On real Atlas this returns an ops counter per index — used to drop indexes that no query ever uses.


## 7. Putting it all together — a unified case dossier query

This is the query that motivates the whole NoSQL design: getting the full case context for a complaint in one round trip.

In [39]:
case = db.customer_cases.find_one({'_id': 'CP0010'})
if case:
    print(f"Case {case['_id']} — {case['complaint_type']} ({case['severity']}, {case['status']})")
    print(f"Customer: {case['customer']['customer_id']} from {case['customer']['home_zone']}")
    if case.get('order'):
        print(f"Order:    {case['order']['order_id']} — {case['order']['service_type']} "
              f"in {case['order']['pickup_zone']}")
    if case.get('delivery'):
        print(f"Delivery: {case['delivery']['delivery_id']} — status {case['delivery']['delivery_status']}")
    print(f"Incidents linked: {len(case['incidents'])}")
    print(f"Thread length:    {len(case['thread'])}")
    if case.get('compensation_amount'):
        print(f"Compensation:     £{case['compensation_amount']:.2f}")

Case CP0010 — DriverBehaviour (Medium, Open)
Customer: C0486 from WEST
Order:    O00417 — Retail in SOUTH
Delivery: DL00939 — status OnTime
Incidents linked: 1
Thread length:    2
Compensation:     £6.32


## 8. Conclusions

1. **Three collections is the right granularity** — case-level (`customer_cases`), session-level (`app_sessions`), and operational-level (`delivery_journeys`). Each matches a real business read pattern.
2. **Master/reference data stays relational.** Customers, drivers, vehicles, hubs change rarely and are accessed by key — embedding them everywhere would create huge update fan-out.
3. **Compound indexes work; single-field on low-selectivity fields don't.** `(pickup_zone, delivery_status)` is the right shape; `delivery_status` alone is not.
4. **`explain()` should be checked for every new query** — it is the only honest way to know whether an index is actually being used.
5. **The 'hidden complaints' problem is now a one-liner** (`db.customer_cases.find({'delivery.delivery_status': 'OnTime'})`) — exactly the integrated operational view the board asked for.